cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -61,6 +63,7 @@
   private final UUIDGenerator _uuidGenerator;
   private final Map<EndPoint, Timer> _successfulRequestExecutionTimer;
   private final long _completedUserTaskRetentionTimeMs;
+  private final Map<EndPoint, UserTaskCleanerInfo> _userTaskCleanerMap;
```

Comment:
Currently we already have a remove completed task mechanism via setting `_completedUserTaskRetentionTimeMs`. Do we want this semantic categorization be an extension to that? If in-memory retention is 24 hours (or days), it might not save us much to have a per-category retention. If the in-memory retention will potentially be weeks or months, then the memory saving is more significant. Just want to understand the purpose and benefit of semantic categorization more.

Reply:
> Currently we already have a remove completed task mechanism via setting `_completedUserTaskRetentionTimeMs`. Do we want this semantic categorization be an extension to that?

In the current behavior, we  have one giant pool to put all requests regardless of which endpoint they come from. We would like to shard the pool into different multiple subpools each representing a different category. Each such category has its own (1) time-based and (2) number-of-item-based retention -- i.e. they will not share the same `24 hours` retention or X-number of items limit.

> If in-memory retention is 24 hours (or days), it might not save us much to have a per-category retention. If the in-memory retention will potentially be weeks or months, then the memory saving is more significant.

The reasoning is to apply selective retention to different requests. For example, I may not be very interested in the last time someone queried the `partition_load` endpoint and the corresponding response (I can get the historical (to some extent) partition load now without the need for keeping the historical response). However, I would be more interested in seeing the request and the corresponding proposals of a recent `remove_broker` operation. But if I put everything under the same pool, then (1) due to less interesting calls and `number-of-item-based retention`, I will lose valuable information (they will be ejected from history) (2) due to the same time-based retention, I cannot keep more important requests selectively -- e.g. one might want to keep the response of `rebalance` longer than `kafka_cluster_state`.

> Just want to understand the purpose and benefit of semantic categorization more.

Hope the explanation above helps!

Reply:
Will having different retention parameters create inconsistency that users need to be aware of? Can the user use request endpoint type filter to find the interesting requests?